esp-hal/src/rng.rs

```diff
@@ -139,3 +139,112 @@ impl rand_core::RngCore for Rng {
         Ok(())
     }
 }
+
+/// True Random Number Generator (TRNG) driver
+///
+/// The `Trng` struct represents a true random number generator that combines
+/// the randomness from the hardware RNG and an ADC. This struct provides
+/// methods to generate random numbers and fill buffers with random bytes.
+/// Due to pulling the entropy source from the ADC, it uses the associated
+/// regiters, so to use TRNG we need to "occupy" the ADC peripheral.
+///
+/// ```no_run
+/// let analog_pin = io.pins.gpio3.into_analog();
+/// let mut adc1_config = AdcConfig::new();
+/// let mut adc1_pin = adc1_config.enable_pin(analog_pin, Attenuation::Attenuation11dB);
+/// let mut adc1 = ADC::<ADC1>::new(peripherals.ADC1, adc1_config);
+/// let pin_value: u16 = nb::block!(adc1.read_oneshot(&mut adc1_pin)).unwrap();
+///
+/// let mut trng = Trng::new(peripherals.RNG, &mut adc1);
+///
+/// let (mut rng, adc1) = trng.downgrade();
+///
+/// // Fill a buffer with random bytes:
+/// let mut buf = [0u8; 16];
+/// println!("Random bytes: {:?}", buf);
+///
+/// println!("Random u32:   {}", rng.random());
+/// let pin_value: u16 = nb::block!(adc1.read_oneshot(&mut adc1_pin)).unwrap();
+/// println!("ADC reading = {}", pin_value);
+/// ```
+#[cfg(not(esp32p4))]
+pub struct Trng<'d> {
+    /// The hardware random number generator instance.
+    pub rng: Rng,
+    /// A mutable reference to the ADC1 instance.
+    adc: &'d mut Adc<'d, crate::peripherals::ADC1>,
+}
+
+#[cfg(not(esp32p4))]
+impl<'d> Trng<'d> {
+    /// Creates a new True Random Number Generator (TRNG) instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `rng` - A peripheral instance implementing the `RNG` trait.
+    /// * `adc` - A mutable reference to an `Adc` instance.
+    ///
+    /// # Returns
+    ///
+    /// Returns a new `Trng` instance.
+    pub fn new(
+        rng: impl Peripheral<P = RNG>,
+        adc: &'d mut Adc<'d, crate::peripherals::ADC1>,
+    ) -> Self {
+        let gen = Rng::new(rng);
+        crate::soc::trng::ensure_randomness();
+        Self { rng: gen, adc }
+    }
+
+    /// Reads currently available `u32` integer from `TRNG`
+    pub fn random(&mut self) -> u32 {
+        self.rng.random()
+    }
+
+    /// Fills the provided buffer with random bytes.
+    pub fn read(&mut self, buffer: &mut [u8]) {
+        self.rng.read(buffer)
+    }
+
+    /// Downgrades the `Trng` instance to a `Rng` instance and releases the
+    /// ADC1. Returns a tuple containing the `Rng` instance and a mutable
+    /// reference to the `Adc`.
+    pub fn downgrade(&mut self) -> (Rng, &'d mut Adc<'_, crate::peripherals::ADC1>) {
```

Comment:
As per the API guidelines, we should use the peripheral ref pattern instead of a release method. We should call `revert` on drop in this case.

Reply:
Done